<div dir=rtl align=center>

<img src='https://upload.wikimedia.org/wikipedia/fa/thumb/a/a9/Sharif_logo.svg/626px-Sharif_logo.svg.png?20110526112825' alt="SUT logo" width=200 height=200 align=center  >
<br>
<font face="B Yekan">
<font color=0F5298 size=7>
یادگیری ماشین<br>
<font color=2565AE size=5>
دانشکده مهندسی صنایع<br>
<font color=2565AE size=4>
دکتر مهدی شریف زاده <br>
<font  size=4>

سعید زواشکیانی
 <br>
بهار 1402<br>

<font color=3C99D size=5>
تمرین عملی 1
<br>
سوال 2
- هزینه سالانه بیمه درمانی
<br>
80 نمره
    
    
____


### Full Name : Parsa Bakhtiari

### Student Number : 97104118

---


<div dir="rtl" style="font-size:16px;">
<font face="B Yekan">
فایل insurance.csv شامل ۶ ویژگی (شامل متغیرها عددی و گروهی (categorical)) از افراد مختلف است که هزینه سالانه بیمه درمانی این افراد (ستون charges بر حسب هزار دلار) در این فایل ثبت شده‌است. هدف، تخمین این هزینه با استفاده از ۶ ویژگی است.

</div>

# Libraries

In [7]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
# add what you want

# Load data

In [8]:
#TODO
df_base = pd.read_csv("insurance.csv")
train_set, test_set = train_test_split(df_base, test_size=0.3, random_state=42)


# A. Averaging (20 points)

<div dir="rtl" style="font-size:16px;">
<font face="B Yekan">

با استفاده از داده‌های آموزش فقط با میانگین‌گیری از خروجی تخمین هزینه را انجام دهید و خطای تست را در این حالت محاسبه کنید.

</div>

In [9]:
#TODO
average_value = np.average(train_set['charges'].values)
test_errors = mean_squared_error(test_set['charges'], np.full((len(test_set['charges'])), average_value))
print(test_errors)

146.75486248250903


# B. Linear regression (40 points)

<div dir="rtl" style="font-size:16px;">
<font face="B Yekan">
یک مدل رگرسیون خطی شامل تمام ویژگی‌ها آموزش دهید. خطای آموزش و تست را گزارش کنید و نتایج را با بخش الف مقایسه کنید. توجه کنید که دادگان حاوی چند ویژگی گروهی هستند که برای استفاده از آن لازم است کدگذاری (encoding) عددی مناسب به کار ببرید. نحوه انجام این کار و نتیجه آن را بیان کنید.
</div>

In [10]:
# TODO
df_reform = pd.get_dummies(
    df_base, columns=['sex', 'smoker', 'region'], drop_first=True)
df_reform.head

<bound method NDFrame.head of       age     bmi  children  charges  sex_male  smoker_yes  region_northwest  \
0      19  27.900         0   16.885         0           1                 0   
1      18  33.770         1    1.726         1           0                 0   
2      28  33.000         3    4.449         1           0                 0   
3      33  22.705         0   21.984         1           0                 1   
4      32  28.880         0    3.867         1           0                 1   
...   ...     ...       ...      ...       ...         ...               ...   
1333   50  30.970         3   10.601         1           0                 1   
1334   18  31.920         0    2.206         0           0                 0   
1335   18  36.850         0    1.630         0           0                 0   
1336   21  25.800         0    2.008         0           0                 0   
1337   61  29.070         0   29.141         0           1                 1   

      reg

In [11]:
train_set_reform, test_set_reform = train_test_split(df_reform, test_size=0.3, random_state=42)
lin_reg = LinearRegression()
lin_reg.fit(train_set_reform.drop('charges', axis=1), train_set_reform['charges'])


LinearRegression()

In [12]:
lin_reg.predict(np.array([[19, 27.900, 0, 0, 1, 0, 0, 1]]))

C:\Users\parsa\AppData\Roaming\Python\Python310\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([25.02727574])

In [13]:
train_errors = mean_squared_error(train_set_reform['charges'], lin_reg.predict(train_set_reform.drop('charges', axis=1)))
test_errors = mean_squared_error(test_set_reform['charges'], lin_reg.predict(test_set_reform.drop('charges', axis=1)))
print(train_errors)
print(test_errors)

37.730673288125395
33.78021054300853


<div dir="rtl" style="font-size:16px;">
<font face="B Yekan">
با الگوریتم one Hot به جای داده های کتگوریکال داده عددی وارد دیتا فریم میکنیم. دلیل این انتخاب کم کردن خطای پیش بینی در داده region بود زیرا بیش از دو مقدار داشت. داده ها استاندارد سازی نشده اند. با توجه به این مسئله مقدار خطای آموزش 33.7 و مقدار خطایی تست 33.8 شده است.
</div>